vowpalwabbit/parser.cc

```diff
@@ -1005,59 +932,51 @@ void finish_example(vw& all, example& ec)
   if (!is_ring_example(all, &ec))
     return;
 
-  mutex_lock(&all.p->output_lock);
-  all.p->local_example_number++;
-  condition_variable_signal(&all.p->output_done);
-  mutex_unlock(&all.p->output_lock);
+  {
+    std::lock_guard<std::mutex> lock(all.p->output_lock);
+    all.p->local_example_number++;
+    all.p->output_done.notify_one();
+  }
 
   clean_example(all, ec, false);
 }
 }
 
 void thread_dispatch(vw& all, v_array<example*> examples)
 {
-  mutex_lock(&all.p->examples_lock);
+  std::lock_guard<std::mutex> lock(all.p->examples_lock);
   all.p->end_parsed_examples+=examples.size();
-  condition_variable_signal_all(&all.p->example_available);
-  mutex_unlock(&all.p->examples_lock);
+  all.p->example_available.notify_all();
 }
 
-#ifdef _WIN32
-DWORD WINAPI main_parse_loop(LPVOID in)
-#else
-void *main_parse_loop(void *in)
-#endif
+void main_parse_loop(vw* all)
 {
-  vw* all = (vw*)in;
   parse_dispatch(*all, thread_dispatch);
-  return 0L;
 }
 
 namespace VW
 {
 example* get_example(parser* p)
 {
-  mutex_lock(&p->examples_lock);
+  std::unique_lock<std::mutex> lock(p->examples_lock);
+
   if (p->end_parsed_examples != p->used_index)
   {
     size_t ring_index = p->used_index++ % p->ring_size;
     if (!(p->examples+ring_index)->in_use)
       cout << "error: example should be in_use " << p->used_index << " " << p->end_parsed_examples << " " << ring_index << endl;
     assert((p->examples+ring_index)->in_use);
-    mutex_unlock(&p->examples_lock);
     return p->examples + ring_index;
   }
   else
   {
     if (!p->done)
     {
-      condition_variable_wait(&p->example_available, &p->examples_lock);
-      mutex_unlock(&p->examples_lock);
+      p->example_available.wait(lock);
       return get_example(p);
```

Comment:
This extends for how long `examples_lock` is held.
What is the locking semantics of  `get_example` ?